### PR TITLE
Refactor DirichletBC

### LINKED ISSUE #507: Allow DirichletBC to be set with a Constant
At present, a DirichletBC can only be set with a `Function`, but it is often just a `Constant` value.

### PR DESCRIPTION
**Pull request by jorgensd**
Superseeds: #1829 and resolves #507 adding `dolfinx.Constant` as a valid input to the DirichletBC.

Refactoring of getting degrees of freedom from remote for topological searches, all different combinations:
`V`, `V.sub(i)`, `(V.sub(i)`, `V.sub(i).collapse())` uses the same functionality to fetch dofs that are not on an owned entity (facet), but is a dof on the process.

### CODE REVIEW

File: cpp/dolfinx/fem/DirichletBC.h [hunk 1]
```diff
@@ -168,6 +170,59 @@ class DirichletBC
...
+  /// @param[in] dofs Degree-of-freedom block indices (@p
+  /// std::vector<std::int32_t>) from the input function space to constrain
+  /// @note The dofs will be unrolled with the dofmap block size
+  template <typename U>
+  DirichletBC(const std::shared_ptr<const fem::Constant<T>>& g, U&& dofs,
+              const std::shared_ptr<const fem::FunctionSpace>& V)
+      : _function_space(V), _g(g), _dofs0(std::forward<U>(dofs))
+  {
+    assert(_function_space);
+    std::visit(
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.h:187]: I don't understand what this is for. The Function interface specifies the type for `g`, so why is there a visitor pattern to check?

File: cpp/dolfinx/fem/DirichletBC.h [hunk 2]
```diff
@@ -254,16 +327,46 @@ class DirichletBC
...
-    xtl::span<const T> g = _g->x()->array();
-    for (std::size_t i = 0; i < _dofs0.size(); ++i)
-    {
-      if (_dofs0[i] < (std::int32_t)x.size())
-      {
-        assert(_dofs1_g[i] < (std::int32_t)g.size());
-        x[_dofs0[i]] = scale * g[_dofs1_g[i]];
-      }
-    }
+    std::visit(
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.h:330]: There are much simpler ways to do this, see https://www.cppstories.com/2018/06/variant/.

File: cpp/dolfinx/fem/DirichletBC.cpp [hunk 3]
```diff
@@ -107,6 +169,8 @@ get_remote_bcs1(const common::IndexMap& map,
   const std::vector<std::int64_t>& ghosts = map.ghosts();
 
   // Build map from ghost to local position
+  // NOTE: Should we use map here or just one vector with ghosts and
```
Comment (IgorBaratta) [inline cpp/dolfinx/fem/DirichletBC.cpp:172]: Sorting takes `O(n * log(n))` , which is the same cost as the creation of the map.
`map.find` and `lower_bound`  are both `O(log(n)` . 

So if the ghosts were sorted I would use a vector intead of a map.

Comment (jorgensd) [reply cpp/dolfinx/fem/DirichletBC.cpp:172]: Currently ghosts are not sorted, so I guess this is a thing we can improve if we change the indexmap, ref #1816

File: cpp/dolfinx/fem/DirichletBC.cpp [hunk 4]
```diff
@@ -50,55 +51,119 @@ dolfinx::MPI::Comm create_symmetric_comm(MPI_Comm comm)
...
+
+  if (bs_map == 1)
+    map.local_to_global(dofs_local, dofs_global);
+  else
+  {
+    // Convert dofs indices to 'block' map indices
+    std::vector<std::int32_t> dofs_local_block;
+    dofs_local_block.reserve(num_dofs_block);
+    for (std::size_t i = 0; i < dofs_local.size(); i += bs)
+      dofs_local_block.push_back(dofs_local[i] / bs_map);
```
Comment (jorgensd) [inline cpp/dolfinx/fem/DirichletBC.cpp:146]: @garth-wells here you have added the assumption that the dofs are sorted and contain no duplicates (which the input currently do) . Should we add that as a `@note for the input arguments`?

Comment (garth-wells) [reply cpp/dolfinx/fem/DirichletBC.cpp:146]: > @garth-wells here you have added the assumption that the dofs are sorted and contain no duplicates 

Not sure that is entirely true on ordering. There is an assumption that blocks remained blocked.

> (which the input currently do) . Should we add that as a `@note for the input arguments`?

The docs already say the dofs must be sorted (in docs for in argument). Feel free to add more if you think it's needed.

Comment (jorgensd) [reply cpp/dolfinx/fem/DirichletBC.cpp:146]: Agree that inside `get_remote_dofs` it does not have to be fully sorted, but the input has to be "blocked".
I will add a note on `get_remote_dofs` that we assume that the input dofs are blocked, but not necessarily sorted.

**Review by garth-wells — APPROVED** [review]

**Pull request merged**